First draft of publish process (#302)
dandi/dandi-api: doc/design/publish-1.md

Only dandiset owners have permission to use the API endpoint.
* During development, only admins can publish.

If a Version is not `VALID` or any of the Assets in the Version is not `VALID`, publish is not allowed.

## Publish process
1. The dandiset is locked so that no other publishes can happen simultaneously.

2. A new published Version is created with the required publish metadata information.
It is initialy empty.

3. A new DOI is created that points to the new Version.
This is an API call to an external service, https://datacite.org/.
This involves digesting the Version metadata into something that can be used by Datacite.
See the doi-generation-1.md design document for more details.

4. For every draft Asset in the draft Version, the required publish metadata fields are injected, turning those Assets into published Assets.
Published Assets, e.g. Asssets that already have those fields from previous publishes, _are skipped_.
Because they describe Assets that have already been published in a past version, there is no reason to re-publish them.

5. Associate all of the publish Assets, new and old, with the new published Version.

> 6. The dandiset is unlocked.

Reviewer: ```suggestion
6. The dandiset is unlocked.

7. An email is sent to the owners that the Dandiset is published.
```

# Moving Parts

> This is a summary of system components that are relevant to publish and how they relate to eachother.

Reviewer 2: ```suggestion
This is a summary of system components that are relevant to publish and how they relate to each other.
```

### Dandiset
A DANDI dataset.

> Each dandiset is guaranteed to have a draft Version, and can also have any number of published Versions.

Reviewer 2: clarification: ATM "is guaranteed to have a draft Version" is true.  In the future we did think about possibility to archive dandisets, so no mutable `draft`  would be left behind, only published Versions.

Author: Lets leave this for now, and clarify with a bigger, better design doc when we care about that feature.

### Version

> Versions consist of a JSON blob of metadata and any number of Assets.

Reviewer 2: ```suggestion
Versions consist of a JSON blob of Dandiset metadata and any number of Assets.
```
right?

There are two kinds of Version: draft and published.
Published Versions are immutable, only draft Versions are editable (except admins, who have superpowers).

> Versions also track their validation status.

Reviewer 2: `### Version` is not under `### Dandiset` so a bit confusing here on which metadata (above suggestion) or validation we are talking about:  validation of dandiset record (metadata), of each asset, or a composite ( everything pertinent to Version: dandiset metadata + all assets metadata, etc).  If I got it right may be:
```suggestion
Versions also track their validation status (comprised of validation of dandiset and assets metadata validation, etc).
```

Author: As currently implemented, Versions only track the validation of their own metadata, not their Assets. You could have a `VALID` version with `INVALID` assets. When a publish is attempted, the `INVALID` assets are checked, so for publishing purposes the version is considered `INVALID`, despite having a column that is marked `VALID`.

```suggestion
Versions also track the validation status of their Dandiset metadata. Asset metadata validation is not recorded in the Version.
```

### AssetBlob

> This is basically just a pointer to a blob in S3.

Reviewer 2: ```suggestion
This is basically just a pointer to a blob in `blobs/` on S3.
```

AssetBlobs are central to how we implement deduplication of blobs in S3.

> Multiple Assets can refer to the same AssetBlob, meaning we only need to store each blob in S3 once.

Reviewer 2: ```suggestion
Multiple Assets can refer to the same AssetBlob, meaning we only need to store each blob in S3 once.
Each Asset could have a different metadata record for the same AssetBlob.
```
if we are to clarify, I think this might be valuable to add

Whenever a new AssetBlob is created, a background task runs to compute the sha256 checksum (a "digest") of the blob.

> This can take a long time for large blobs, as it must stream all of the data from S3.

Reviewer 2: ```suggestion
This can take a long time for large blobs, as it must stream all of the data from S3.
Computed sha256 checksum is associated with the blob DB record (so no Asset metadata record is changed).
```

### Asset
Assets consist of a JSON blob of metadata and a reference to an AssetBlob.
Versions and Assets are many-to-many; a Version can have any number of Assets, and an Asset can belong to any number of Versions (even, hypothetically, versions in different Dandisets).
Assets also track their validation status.

> Similarly to Versions, there are two kinds of Asset: draft and published.
> Draft Assets are only allowed to belong to draft Versions.
> Published Assets are basically the same as draft Assets, but have some extra metadata included that was defined during the publish process (`datePublished`, `publishedBy`, etc).
> 
> Note that published Assets can belong to both draft and published Versions.
> 
> Uploading new data or modifying the metadata of an Asset doesn't actually modify the Asset.
> Instead, it creates a new Asset and replaces the reference to that Asset in the Version.
> This ensures that other Versions which might refer to the original Asset are unmodified, while the Version containing the Asset being modified reflects the changes correctly.

Reviewer 2: I would reorder these two paragraphs to go from basic "Asset" behavior to IMHO yet to be finalized "published vs draft" Asset fresh discussion:
```suggestion
Uploading new data or modifying the metadata of an Asset doesn't actually modify the Asset.
Instead, it creates a new Asset and replaces the reference to that Asset in the Version.
This ensures that other Versions which might refer to the original Asset are unmodified, while the Version containing the Asset being modified reflects the changes correctly.

Similarly to Versions, there are two kinds of Asset: draft and published.
Draft Assets are only allowed to belong to draft Versions.
Published Assets are basically the same as draft Assets, but have some extra metadata included that was defined during the publish process (`datePublished`, `publishedBy`, etc).

Note that published Assets can belong to both draft and published Versions.
```

When a published Asset is modified in this way, the publish metadata is stripped out, effectively creating a new draft Asset.

_In general_, Assets should be immutable; any modification will instead create a new draft Asset.
* One exception to this is calculating the AssetBlob digest.
When the digest completes, it will be injected directly into the metadata of any Assets referencing that AssetBlob.
* The other exception is creating a published Asset during the publish process.

> This involves directly injecting the new metadata fields into the draft Asset, effectively turning it into a published Asset in place.

Reviewer 2: per my question and brief discussion, I do not think we should bother with mutating Assets in DB, and can keep the notion of draft vs published asset metadata just implemented while returning asset record from an endpoint for dandiset_version.  If `version != "draft"` (thus published) -- we populate fields relevant to Published version.  If `draft` - we don't. Then we would stay consistent with the notion of immutable "Asset metadata as provided by user" in DB.
The same goes for 

> When the digest completes, it will be injected directly into the metadata of any Assets referencing that AssetBlob.

and we do not inject anything into actual asset metadata record but just populate it "on the fly" from blob DB record.
And we avoid the need for "the publish metadata is stripped out" which might get hairy IMHO.

If there is a desire to use `contentUrl` as from the first published version for an asset, could be achieved at DB query level by searching asset in Versions with minimal non-draft version.  Or we just KISS it and provide version specific contentUrl regardless.

### Metadata Schemas
Version metadata and Asset metadata must conform to defined schemas.
There are two sets of schemas.
The first applies to both draft and published metadata, and is used by the web metadata editor.
The second is more restrictive, and only applies to published metadata.
It requires additional fields like `datePublished`, `publishedBy`, and `doi` that are only knowable at publish time.
The second one is used to validate metadata prior to publish.

The JSON schemas can be found at https://github.com/dandi/schema.

The pydantic models can be found and imported from https://github.com/dandi/dandischema.

> dandi-api will use the pydantic models for validation.

Reviewer 2: At least ATM we have no provisioning for validating against different schema versions via pydantic and I do not see yet how we could in the long(er) run due to all possible dependencies on pydantic version/logic etc.  Multiple strategies could be explored (e.g. "upgrade provided metadata to the most recent pydantic model version and validate") but we do not yet have support for that even designed.  Filed https://github.com/dandi/dandischema/issues/5 .

In the short term, I guess we can go "pydantic model" way but would need to orchestrate reextraction (for assets) and upgrade (for dandisets) of all metadata records and make publish accept only the most recent one. Thus here we would need to abstract away IMHO into

```suggestion
dandi-api will use the `dandischema` validation API (ATM will not support older schema versions and thus require metadata re-extraction/upgrade).
```
or smth like that.

Author: My understanding was that it was feasible to write a library that would return the correct pydantic models for a given version. 

```suggestion
dandi-api will look up the JSON schema for the specified `schemaVersion` and use it for validation. Hopefully at some point we can find a way to provide arbitrary pydantic model versions and use those instead.
```

# Pre-publish

## Uploading data / Modifying metadata

> Whenever changes are made to the draft Version of a Dandiset, metadata will almost always be modified.

Reviewer 2: I guess I am confused by lack of explicit difference between "metadata JSON blob" (as in DB) and "metadata record returned by API" (JSOB blob + computed).

Author: In my draft, there is no difference. IMHO that is one of the benefits of not injecting information into the metadata.

* During development, only admins can publish.

If a Version is not `VALID` or any of the Assets in the Version is not `VALID`, publish is not allowed.

## Publish process
1. The dandiset is locked so that no other publishes can happen simultaneously.

2. A new published Version is created with the required publish metadata information.
It is initialy empty.

3. A new DOI is created that points to the new Version.
This is an API call to an external service, https://datacite.org/.
This involves digesting the Version metadata into something that can be used by Datacite.
See the doi-generation-1.md design document for more details.

4. For every draft Asset in the draft Version, the required publish metadata fields are injected, turning those Assets into published Assets.
Published Assets, e.g. Asssets that already have those fields from previous publishes, _are skipped_.
Because they describe Assets that have already been published in a past version, there is no reason to re-publish them.

5. Associate all of the publish Assets, new and old, with the new published Version.

6. The dandiset is unlocked.

7. An email is sent to the owners that the Dandiset is published.

> 7. Asynchronously, manifest files (`dandiset.yaml` and `assets.yaml`) are written out to S3 describing the newly published Version by a background task.

Author: ```suggestion
8. Asynchronously, manifest files (`dandiset.yaml` and `assets.yaml`) are written out to S3 describing the newly published Version by a background task.
```
